modules/core/src/experimental/utils/gpu-grid-aggregation/gpu-grid-aggregator.js

```diff
@@ -76,83 +121,452 @@ export default class GPUGridAggregator {
+          fillValue = -Infinity;
+          break;
+        default:
+          // Not a valid operation enum.
+          assert(false);
+          break;
+      }
+
+      const aggregationSize = numCol * numRow * ELEMENTCOUNT;
+      aggregationData = getFloatArray(aggregationData, aggregationSize, fillValue);
+      if (combineMaxMin) {
+        maxMinData = getFloatArray(maxMinData, ELEMENTCOUNT);
+        // RGB for max value
+        maxMinData.fill(-Infinity, 0, ELEMENTCOUNT - 1);
+        // Alpha for min value
+        maxMinData[ELEMENTCOUNT - 1] = Infinity;
+      } else {
+        // RGB for min/max values
+        // Alpha for total count
+        if (needMin) {
+          minData = getFloatArray(minData, ELEMENTCOUNT, Infinity);
+          minData[ELEMENTCOUNT - 1] = 0;
+        }
+        if (needMax) {
+          maxData = getFloatArray(maxData, ELEMENTCOUNT, -Infinity);
+          maxData[ELEMENTCOUNT - 1] = 0;
+        }
+      }
+      results[id] = Object.assign({}, weights[id], {
+        aggregationData,
+        minData,
+        maxData,
+        maxMinData
+      });
+    }
+    return results;
+  }
+
+  /* eslint-disable max-statements */
+  runAggregationOnCPU(opts) {
+    const {positions, cellSize, gridTransformMatrix, viewport, projectPoints} = opts;
+    let {weights} = opts;
+    const {numCol, numRow} = this.state;
+    const results = this.initCPUResults(opts);
+    // screen space or world space projection required
+    const gridTransformRequired = this.shouldTransformToGrid(opts);
+    let gridPositions = [];
+
+    assert(gridTransformRequired || opts.changeFlags.cellSizeChanged);
+
+    let posCount;
+    if (gridTransformRequired) {
+      this.setState({gridPositions});
+      posCount = positions.length / 2;
+    } else {
+      gridPositions = this.state.gridPositions;
+      weights = this.state.weights;
+      posCount = gridPositions.length / 2;
+    }
+
+    const validCellIndices = new Set();
+    for (let posIndex = 0; posIndex < posCount; posIndex++) {
+      let gridPos;
+      if (gridTransformRequired) {
+        const pos = [positions[posIndex * 2], positions[posIndex * 2 + 1]];
+        if (projectPoints) {
+          gridPos = viewport.project([pos[0], pos[1]]);
+        } else {
+          gridPos = worldToPixels([pos[0], pos[1], 0], gridTransformMatrix).slice(0, 2);
+        }
+        gridPositions.push(...gridPos);
+      } else {
+        gridPos = [gridPositions[posIndex * 2], gridPositions[posIndex * 2 + 1]];
+      }
+
+      const x = gridPos[0];
+      const y = gridPos[1];
+      const colId = Math.floor(x / cellSize[0]);
+      const rowId = Math.floor(y / cellSize[1]);
+      if (colId >= 0 && colId < numCol && rowId >= 0 && rowId < numRow) {
+        const cellIndex = (colId + rowId * numCol) * ELEMENTCOUNT;
+        validCellIndices.add(cellIndex);
+        this.calculateAggregationData({weights, results, cellIndex, posIndex});
+      }
+    }
+
```

Comment:
I might be missing something completely, but I can't figure out how the mean value is calculated.  As I read it, this.calculateAggregationData sums up values for each position and get the count when the mean operation is set. However, I don't see where the sum for each cell is divided by the count.

Reply:
@richardowright , I will be adding `mean` calculation for CPU and GPU paths in a separate PR given the amount of changes done in this PR already.  Will be tracked in #2291 and will do that right after this one is landed.